src/configuration.ts

```diff
@@ -1,180 +1,674 @@
+        pj = require(`${appRoot.path}/package.json`);
+    } catch (e) {
+        logger.warn(`failed to load package.json: ${e.message}`);
+    }
+    pj.name = pj.name || "atm-client-" + guid();
+    pj.version = pj.version || "0.0.0";
+    pj.keywords = pj.keywords || [];
+
+    const cfg: Configuration = {
+        name: pj.name,
+        version: pj.version,
+        keywords: pj.keywords,
+        teamIds: [],
+        groups: [],
+        environment: "local",
+        application: pj.name.replace(/^@.*?\//, ""),
+        policy: "ephemeral",
+        endpoints: {
+            api: "https://automation.atomist.com/registration",
+            graphql: "https://automation.atomist.com/graphql/team",
+        },
+        http: {
+            enabled: true,
+            host: "localhost",
+            port: 2866,
+            auth: {
+                basic: {
+                    enabled: true,
+                },
+                bearer: {
+                    enabled: true,
+                },
+            },
+            customizers: [],
+        },
+        ws: {
+            enabled: true,
+            termination: {
+                graceful: false,
+            },
+            compress: false,
+        },
+        cluster: {
+            enabled: false,
+        },
+        logging: {
+            level: "info",
+        },
+        applicationEvents: {
+            enabled: false,
+        },
+        commands: [],
+        events: [],
+        ingesters: [],
+        listeners: [],
+        initializers: [],
+    };
+    return cfg;
+}
+
+/**
+ * Return Atomist user configuration directory.
+ */
+function userConfigDir(): string {
+    const home = process.env[process.platform === "win32" ? "USERPROFILE" : "HOME"];
+    return path.join(home, ".atomist");
+}
+
+/**
+ * Return user automation client configuration path.
+ */
+function userConfigPath(): string {
+    const clientConfigFile = "client.config.json";
+    return path.join(userConfigDir(), clientConfigFile);
 }
 
 /**
  * Write user config securely, creating directories as necessary.
  */
 export function writeUserConfig(cfg: UserConfig): Promise<void> {
-    return fs.ensureDir(UserConfigDir)
-        .then(() => fs.chmod(UserConfigDir, 0o700))
-        .then(() => fs.writeJson(UserConfigFile, cfg, {
+    const cfgDir = userConfigDir();
+    return fs.ensureDir(cfgDir)
+        .then(() => fs.chmod(cfgDir, 0o700))
+        .then(() => fs.writeJson(userConfigPath(), cfg, {
             spaces: 2,
-            encoding: UserConfigEncoding,
+            encoding: "utf8",
             mode: 0o600,
         }));
 }
 
 /**
- * Read and return user config.
+ * Read and return user config from UserConfigFile.
  */
 export function getUserConfig(): UserConfig {
-    try {
-        if (fs.existsSync(UserConfigFile)) {
-            return fs.readJsonSync(UserConfigFile) as UserConfig;
+    let cfg: UserConfig;
+    if (fs.existsSync(userConfigPath())) {
```

Comment:
If the file is not there the try will properly handle and report that case, as well as if it is not readable, not proper JSON, etc. It also avoids a race condition.

From https://nodejs.org/api/fs.html#fs_fs_exists_path_callback:

> Using fs.exists() to check for the existence of a file before calling fs.open(), fs.readFile() or fs.writeFile() is not recommended. Doing so introduces a race condition, since other processes may change the file's state between the two calls. Instead, user code should open/read/write the file directly and handle the error raised if the file does not exist.

Reply:
I want to avoid even attempting to load the file and not report the fact it is missing. IMHO it looks bad if the first thing we log when the client starts is a warning about a missing file. 

That’s why I think using exists to check if the file is there is ok. If we ever hit a race condition that suddenly the file goes missing, this is still being handled in the inner catch.